Plutarch/Core.hs

```diff
@@ -306,101 +307,21 @@ class PDSL edsl => PAp (f :: Type -> Type) edsl where
--- TODO: Possibly show type in question?
-instance
-  {-# OVERLAPPABLE #-}
-  ( TypeError
-      ( Text "Can not embed type that contains: "
-          :$$: ShowType a
-      )
-  , PIsProduct edsl as
-  ) =>
-  PIsProduct edsl (a : as)
-  where
-  eisProduct edsl _ = error "unreachable" $ eisProduct edsl (Proxy @as)
-
-instance (IsPType edsl a, PIsProduct edsl as) => PIsProduct edsl (Term edsl a : as) where
-  eisProduct edsl _ =
-    let prev = eisProduct edsl (Proxy @as)
-     in case prev of
-          PIsProductR {inner = _ :: Proxy asi, to, from} ->
-            PIsProductR
-              { inner = Proxy @(a : asi)
-              , to = \(x SOP.:* xs) -> SOP.I x SOP.:* to xs
-              , from = \(SOP.I x SOP.:* xs) -> x SOP.:* from xs
-              }
-
-data PIsSumR (edsl :: PDSLKind) (a :: [[Type]]) = forall inner.
-  SOP.All2 (IsPType edsl) inner =>
-  PIsSumR
-  { inner :: Proxy inner
-  , to :: SOP.SOP (Term edsl) inner -> SOP.SOP SOP.I a
-  , from :: SOP.SOP SOP.I a -> SOP.SOP (Term edsl) inner
-  }
-
-class PIsSum (edsl :: PDSLKind) (a :: [[Type]]) where
-  eisSum :: Proxy edsl -> Proxy a -> PIsSumR edsl a
-
-instance PIsSum edsl '[] where
-  eisSum _ _ =
-    PIsSumR
-      { inner = Proxy @'[]
-      , to = \case {}
-      , from = \case {}
-      }
-
-instance (PIsProduct edsl a, PIsSum edsl as) => PIsSum edsl (a : as) where
-  eisSum edsl _ =
-    case eisProduct edsl (Proxy @a) of
-      PIsProductR {inner = _ :: Proxy innerh, to = toh, from = fromh} ->
-        case eisSum edsl (Proxy @as) of
-          PIsSumR {inner = _ :: Proxy innert, to = tot, from = fromt} ->
-            PIsSumR
-              { inner = Proxy @(innerh : innert)
-              , to = \case
-                  SOP.SOP (SOP.Z x) -> SOP.SOP $ SOP.Z $ toh x
-                  SOP.SOP (SOP.S x) -> case tot $ SOP.SOP $ x of SOP.SOP y -> SOP.SOP (SOP.S y)
-              , from = \case
-                  SOP.SOP (SOP.Z x) -> SOP.SOP $ SOP.Z $ fromh x
-                  SOP.SOP (SOP.S x) -> case fromt $ SOP.SOP $ x of SOP.SOP y -> SOP.SOP (SOP.S y)
-              }
+class (forall a. t ~ Term edsl a) => PIsTerm edsl t
+instance (forall a. t ~ Term edsl a) => PIsTerm edsl t
+
+class (SOP.All (PIsTerm edsl) as) => PIsProduct (edsl :: PDSLKind) (as :: [Type])
+instance (SOP.All (PIsTerm edsl) as) => PIsProduct (edsl :: PDSLKind) (as :: [Type])
+
+class (SOP.All2 (PIsTerm edsl) as) => PIsSum (edsl :: PDSLKind) (as :: [[Type]])
+instance (SOP.All2 (PIsTerm edsl) as) => PIsSum (edsl :: PDSLKind) (as :: [[Type]])
 
 class
   ( PGeneric a
-  , PIsSum edsl (SOPG.GCode (PConcrete edsl a))
-  , PReprSort a ~ PReprSOP
-  ) =>
-  PIsSOP (edsl :: PDSLKind) (a :: PType)
-  where
-  esop :: Proxy edsl -> Proxy a -> PIsSumR edsl (SOPG.GCode (PConcrete edsl a))
-instance
-  ( PGeneric a
-  , PIsSum edsl (SOPG.GCode (PConcrete edsl a))
+  , PIsSum edsl (SOP.Code (PConcrete edsl a))
```

Comment:
This seems like a mistaken change? You probably still want `GCode`.

Reply:
`SOP.Code` is a synonym. Nevertheless, using in place of `GCode` has caused me problems, so I'm adding it back.

Reply:
`SOP.Code` is not a synonym.

Reply:
😮 This led me astray: https://hackage.haskell.org/package/generics-sop-0.5.1.2/docs/Generics-SOP.html#t:Code:~:text=type%20Code%20a%20%3D%20GCode%20a